Optimize sort on numeric long and date fields

mayya-sharipova opened the pull request:
Rewrite sort on numeric long or date field, 
as `LongPoint.newDistanceFeatureQuery`.
This allow to significantly speed up the sorting.

The rewritten query will be be a bool query,
with `FILTER` clause on the original query,
and `SHOULD` clause on LongDistanceFeatureQuery.

closes #37043

elasticmachine commented:
Pinging @elastic/es-search

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+
+        // 2. Iterate over DocValues and fill fieldDocs
+        Long missingValue = (Long) sortFields[0].getMissingValue(); // for rewritten numeric sort, when we had a single sort on Long field
+        SortedNumericDocValues sdvs = MultiDocValues.getSortedNumericValues(reader, sortFields[0].getField());
```
@jpountz Is it fine to use `MultiDocValues.getSortedNumericValues` or is there is some other way to get values for docs?

jimczi replied:
I wonder if this could be simplified if we add the numeric field as the secondary sort. This way you can access the values directly in the returned field docs ?

mayya-sharipova commented:
@jpountz Adrien, I have finished the implementation of hits optimization on sort, but would like to get your feedback before doing further work.

The main question is what to do with multiple values per document? Two challenges here:
1) `LongDistanceFeatureQuery` has[ its own way](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/document/LongDistanceFeatureQuery.java#L118-L141) to select a value, which can be different from what a user specifies in sort mode (min, max, avg, median)
2) Difficulty in filling a proper value  when [covering from topDocs to fieldDocs](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/document/LongDistanceFeatureQuery.java#L118-L141)

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+        // check if this is a field of type Long, that is indexed and has doc values
+        String fieldName = sortField.getField();
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return  null;
```
if the field isn't mapped, we could maybe return eg. a `MatchAllDocsQuery`?

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+        String fieldName = sortField.getField();
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return  null;
+        if (fieldType.typeName() != "long") return null;
```
Can you use `.equals` instead? I don't we would get an interned string all the time, so it shouldn't really matter, but I'd rather like to avoid relying on it.

jimczi replied:
I guess it's because it's a work in progress but the optimization can work on any numeric and date field ?

mayya-sharipova replied:
@jimczi The optimization only works for the numeric long field and date field.  Date field was added as well

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+        if (fieldType.hasDocValues() == false) return null;
+        byte[] minValueBytes =  PointValues.getMinPackedValue(reader, fieldName);
+        byte[] maxValueBytes =  PointValues.getMaxPackedValue(reader, fieldName);
+        if ((maxValueBytes == null) || (minValueBytes == null)) return null; // no values on this shard
```
maybe return a MatchAllDocsQuery in that case

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return  null;
+        if (fieldType.typeName() != "long") return null;
+        if (fieldType.indexOptions() == IndexOptions.NONE) return null;
```
this is always going to be NONE on long fields, which are indexed with points, you should check the point dimension count instead

mayya-sharipova replied:
@jpountz  Actually for long fields, for `fieldType.indexOptions()` I am getting `IndexOptions.DOCS_AND_FREQS_AND_POSITIONS`  and for `fieldType.pointDataDimensionCount() ` I am getting 0.  Should we modify first `NumberFieldType` to have IndexOptions.NONE  and `dataDimensionCount = 1`?

jimczi replied:
+1 to set the dimension count on numeric and date fields.

jpountz replied:
oh sorry I got confused between what the actual fieldtype would look like at the Lucene level and what the Elasticsearch fieldtype looks like. Changing the MappedFieldType would be nice but from what I remember it might not be straightforward due to the fact that we have a fair number of functions that assume that indexOptions != NONE means "indexed". So let's not try to do it as part of this PR.

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+        byte[] minValueBytes =  PointValues.getMinPackedValue(reader, fieldName);
+        byte[] maxValueBytes =  PointValues.getMaxPackedValue(reader, fieldName);
+        if ((maxValueBytes == null) || (minValueBytes == null)) return null; // no values on this shard
+
```
I think we need to check the missing value as well to make sure this optimization is applicable.

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+        long minValue = LongPoint.decodeDimension(minValueBytes, 0);
+        long maxValue = LongPoint.decodeDimension(maxValueBytes, 0);
+        final long origin = (sortField.getReverse()) ? maxValue : minValue;
+        final long pivotDistance = maxValue == minValue ? maxValue/2 : (maxValue - minValue)/2;
```
Maybe `minValue == maxValue` is a case when we should return a DocValueFieldExistsQuery since there is no point in computing distances in such a case.

jpountz commented:
> Detect that documents have multiple values and don't run optimizations in this case

+1 to this approach. You can do that easily by comparing PointValues#size and docCount. I like the 2nd approach less since it is less transparent.

jimczi commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+        Sort sort = searchContext.sort().sort;
+        if (Sort.RELEVANCE.equals(sort)) return null;
+        if (Sort.INDEXORDER.equals(sort)) return null;
+        if (sort.getSort().length > 1) return null; // we need only a single sort
```
I don't think this should prevent the optimization to kick in. The only condition is that the primary sort is performed on a numeric field indexed with points and doc_values no matter what fields is used for tiebreaking.

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -177,6 +187,38 @@ static boolean execute(SearchContext searchContext,
+                        }
+                        // add the numeric field as the last sort, so we can access its values later
+                        newSortFields[oldSortFields.length] = oldSortFields[0];
+                        newFormats[oldSortFields.length] = oldFormats[0];
```
Actually we should make it second, not last. In some cases, different values will be mapped to the same score because floats don't have infinite accuracy. So we should tie-break on the field that we are sorting on.

mayya-sharipova replied:
@jpountz  thanks Adrien, corrected!

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -177,6 +187,35 @@ static boolean execute(SearchContext searchContext,
+                        for (int i = 0; i < oldSortFields.length; i++) {
+                            newSortFields[i + 1] = oldSortFields[i];
+                            newFormats[i + 1] = oldFormats[i];
+                        }
```
let's replace this for-loop with a call to System#arraycopy?

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +349,67 @@ static boolean execute(SearchContext searchContext,
+        if (searchContext.mapperService() == null) return null; // mapperService can be null in tests
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return null; // for unmapped fields, default behaviour depending on "unmapped_type" flag
+         if ((fieldType.typeName().equals("long") == false) && (fieldType instanceof DateFieldType == false)) return null;
```
extra indentation?

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +349,67 @@ static boolean execute(SearchContext searchContext,
 
+     private static Query tryRewriteNumericLongOrDateSort(SearchContext searchContext, IndexReader reader) throws IOException {
+        // child docs can inherit scores from parent, so avoid has_parent query
+        if (searchContext.request().source().query().getName().equals("has_parent")) return null;
```
can you explain this one, I don't understand why this is required.

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +349,67 @@ static boolean execute(SearchContext searchContext,
+        if (searchContext.collapse() != null) return null;
+        Sort sort = searchContext.sort().sort;
+        if (Sort.RELEVANCE.equals(sort)) return null;
+        if (Sort.INDEXORDER.equals(sort)) return null;
```
I think those two cases would be covered below by the `fieldName == null` check?

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +349,67 @@ static boolean execute(SearchContext searchContext,
+            Object[] newFieldValues = new Object[oldFieldValues.length - 1];
+            for (int i = 0; i < oldFieldValues.length - 1; i++) {
+                newFieldValues[i] = oldFieldValues[i + 1];
+            }
```
Use Arrays#copyOfRange?

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +349,67 @@ static boolean execute(SearchContext searchContext,
+
+        byte[] minValueBytes = PointValues.getMinPackedValue(reader, fieldName);
+        byte[] maxValueBytes = PointValues.getMaxPackedValue(reader, fieldName);
+        if ((maxValueBytes == null) || (minValueBytes == null)) return new MatchAllDocsQuery(); // no values on this shard
```
Apologies for the back-and-forth, I just realized that returning a query that produces constant-scores would _not_ help because when there are multiple sorting criteria, we can only ask hits to have a score that is greater than *or equal to* to best k-th score, rather than strictly greater to like when sorting only by score.

mayya-sharipova commented:
@jpountz  Adrien, thanks for the review. I have tried to address your comments, and this PR is ready for another round of review.

jpountz commented:
Something I meant to ask as well: have you tried it out on some dataset, does it make sorting faster?

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +348,76 @@ static boolean execute(SearchContext searchContext,
+
+        // check if this is a field of type Long or Date, that is indexed and has doc values
+        String fieldName = sortField.getField();
+        if (fieldName == null) return null; // happens when _score or _doc is the 1st sort field
```
we should probably also return null when sortField.getType is not long. Not doing it would not cause bugs to my knowledge, but Lucene supports providing custom comparators, while we are assuming the natural number ordering here.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -303,6 +348,76 @@ static boolean execute(SearchContext searchContext,
+        if (searchContext.request() != null && searchContext.request().source().query() != null) {
+            if (searchContext.request().source().query().getName().equals("has_parent"))
+                return null;
+        }
```
@jpountz Thanks I will remove this condition.
The test that previously was failing for me was the last test in `ChildQuerySearchIT::testScoreForParentChildQueriesWithFunctionScore`. 

But after I added another check that there is NO _score sort field among any sort fields, this test doesn't fail anymore.

`has_parent` query  with the  parameter `"score" = true` produces child docs with scores from their parents even if there is a sort field for the child docs.  But this apparently only happens when there is a secondary sort field on `_score` as in the test `.addSort(SortBuilders.fieldSort("c_field3")).addSort(SortBuilders.scoreSort())`
Thus the check that there is not `_score` sort field is enough to ensure we don't run sort optimizations for this kind of queries.

mayya-sharipova commented:
@jpountz Thanks Adrien for the feedback. I have addressed your comments in the last commit.  One thing still left is to do an evaluation on a dataset. I will do the evaluation, and will get back with results.

mayya-sharipova commented:
I ran a first round of rally benchmarks using rally's `geonames` track and sorting on `population` field:
```js
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "population": "desc"
    }
  ]
}
```
with `"warmup-iterations": 200`

The results without optimization:
```
|               Min Throughput | long_sort_population |       1.5 |  ops/s |
|            Median Throughput | long_sort_population |      1.51 |  ops/s |
|               Max Throughput | long_sort_population |      1.51 |  ops/s |
|      50th percentile latency | long_sort_population |   56.7625 |     ms |
|      90th percentile latency | long_sort_population |   62.6092 |     ms |
|      99th percentile latency | long_sort_population |   70.5794 |     ms |
|     100th percentile latency | long_sort_population |   76.6261 |     ms |
| 50th percentile service time | long_sort_population |   54.0481 |     ms |
| 90th percentile service time | long_sort_population |   59.8921 |     ms |
| 99th percentile service time | long_sort_population |   66.1887 |     ms |
|100th percentile service time | long_sort_population |    74.764 |     ms |
```
The results with optimization:
```
|               Min Throughput | long_sort_population |       1.5 |  ops/s |
|            Median Throughput | long_sort_population |      1.51 |  ops/s |
|               Max Throughput | long_sort_population |      1.51 |  ops/s |
|      50th percentile latency | long_sort_population |   62.8559 |     ms |
|      90th percentile latency | long_sort_population |    66.669 |     ms |
|      99th percentile latency | long_sort_population |   73.1161 |     ms |
|     100th percentile latency | long_sort_population |     73.61 |     ms |
| 50th percentile service time | long_sort_population |   59.2964 |     ms |
| 90th percentile service time | long_sort_population |   62.2351 |     ms |
| 99th percentile service time | long_sort_population |   67.9318 |     ms |
|100th percentile service time | long_sort_population |   68.2225 |     ms |
```

`/rally compare --baseline=20190523T160203Z --contender=20190523T161815Z`
<img width="708" alt="Screenshot 2019-05-23 12 57 19" src="https://user-images.githubusercontent.com/5738841/58271557-4fbfa880-7d5a-11e9-81db-dd33f08005f2.png">

Looks like the optimization brought slight performance degradation, but for the longest running queries (in 99-100% percentile), there seems to be an improvement.

mayya-sharipova commented:
Here are results on running  on the `geonames` rally's track.
Thanks @jimczi  for corrections and help.

Running with an operation:
```js
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "population": "desc"
    }
  ]
}
```

Results without optimization:
```
|       50th percentile latency | long_sort_population |   58.0295 |     ms |
|       90th percentile latency | long_sort_population |    62.525 |     ms |
|       99th percentile latency | long_sort_population |   68.5235 |     ms |
|      100th percentile latency | long_sort_population |    72.394 |     ms |
|  50th percentile service time | long_sort_population |   54.1314 |     ms |
|  90th percentile service time | long_sort_population |   57.6252 |     ms |
|  99th percentile service time | long_sort_population |   63.1359 |     ms |
| 100th percentile service time | long_sort_population |   67.0069 |     ms |
```


Results with optimization:
```
|        50th percentile latency | long_sort_population |   7.19378 |     ms |
|        90th percentile latency | long_sort_population |   9.63573 |     ms |
|        99th percentile latency | long_sort_population |   15.9834 |     ms |
|       100th percentile latency | long_sort_population |   17.1643 |     ms |
|   50th percentile service time | long_sort_population |   4.39315 |     ms |
|   90th percentile service time | long_sort_population |   4.75856 |     ms |
|   99th percentile service time | long_sort_population |   10.8195 |     ms |
|  100th percentile service time | long_sort_population |   13.7886 |     ms |
```

Optimization clearly demonstrates speedups:
<img width="724" alt="Screenshot 2019-05-23 17 50 51" src="https://user-images.githubusercontent.com/5738841/58289088-61697600-7d83-11e9-9b94-bf9ab126de90.png">

queries with optimizations ran 5-7 times faster.

jimczi commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -306,6 +352,73 @@ static boolean execute(SearchContext searchContext,
+        if (searchContext.trackScores()) return null;
+        Sort sort = searchContext.sort().sort;
+        SortField sortField = sort.getSort()[0];
+        if (SortField.Type.LONG.equals(IndexSortConfig.getSortFieldType(sortField)) == false) return null;
```
As discussed offline this should take into account `SortedNumericSortField`, see `IndexSortConfig#getSortFieldType` for an example.

jimczi commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -306,6 +352,73 @@ static boolean execute(SearchContext searchContext,
         }
     }
 
+     private static Query tryRewriteNumericLongOrDateSort(SearchContext searchContext, IndexReader reader) throws IOException {
```
We should not activate the optim if trackTotalHits is true or set to Integer.MAX_VALUE since we cannot skip hits in this case. I think that moving these checks to `TopDocsCollectorContext#SimpleTopDocsCollectorContext` would be easier since you can find the final value of `trackTotalHits` easily. We should also not activate this optim if aggregations are part of the query because we don't apply max-score in this case either.

jimczi commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -306,6 +352,73 @@ static boolean execute(SearchContext searchContext,
+        if (missingValuesAccordingToSort == false) return null;
+
+        // check for multiple values
+        if (PointValues.size(reader, fieldName) != PointValues.getDocCount(reader, fieldName)) return null; //TODO: handle multiple values
```
This should work if the multi-valued sort mode is set to `max` when you sort in reverse order (that's the default) and `min` if you sort in natural order (that's the default too). You can check the `SortedNumericSortField` to find these values.

jpountz replied:
This change is already large, maybe we should keep making this optimization work for multi-valued fields as a follow-up?

jimczi replied:
Sure

mayya-sharipova commented:
Here are the results of running of the geonames rally's track on sort ascending:
```js
{
      "name": "long_sort_population",
      "operation-type": "search",
      "body": {
        "query": {
          "match_all": {}
        },
        "sort" : [
          {"population" : "asc"}
        ]
      }
    }
```

Without optimization:
```
|       50th percentile latency | long_sort_population |   56.2071 |     ms |
|       90th percentile latency | long_sort_population |   60.0372 |     ms |
|       99th percentile latency | long_sort_population |   66.0915 |     ms |
|      100th percentile latency | long_sort_population |   67.7832 |     ms |
|  50th percentile service time | long_sort_population |   52.0731 |     ms |
|  90th percentile service time | long_sort_population |   55.3643 |     ms |
|  99th percentile service time | long_sort_population |   60.7257 |     ms |
| 100th percentile service time | long_sort_population |   65.8864 |     ms |
```
with optimization:
```
|       50th percentile latency | long_sort_population |   175.018 |     ms |
|       90th percentile latency | long_sort_population |   184.525 |     ms |
|       99th percentile latency | long_sort_population |   206.141 |     ms |
|      100th percentile latency | long_sort_population |   210.183 |     ms |
|  50th percentile service time | long_sort_population |   171.161 |     ms |
|  90th percentile service time | long_sort_population |   179.845 |     ms |
|  99th percentile service time | long_sort_population |   205.672 |     ms |
| 100th percentile service time | long_sort_population |   209.223 |     ms |
```
<img width="726" alt="Screenshot 2019-05-24 18 47 51" src="https://user-images.githubusercontent.com/5738841/58360039-7b26be00-7e54-11e9-9ea3-443551e6085c.png">

Looks like there is 3-4 times degradation in the performance

mayya-sharipova commented:
Profile from one shard `asc`
```js
"searches": [
    {
        "query": [
            {
                "type": "BooleanQuery",
                "description": "#*:* LongDistanceFeatureQuery(field=,origin=0,pivotDistance=787892250)",
                "time_in_nanos": 1595958190,
                "breakdown": {
                    "set_min_competitive_score_count": 100,
                    "match_count": 0,
                    "shallow_advance_count": 0,
                    "set_min_competitive_score": 231962,
                    "next_doc": 425294028,
                    "match": 0,
                    "next_doc_count": 2279537,
                    "score_count": 2279537,
                    "compute_max_score_count": 0,
                    "compute_max_score": 0,
                    "advance": 30213,
                    "advance_count": 21,
                    "score": 1165159282,
                    "build_scorer_count": 42,
                    "create_weight": 33343,
                    "shallow_advance": 0,
                    "create_weight_count": 1,
                    "build_scorer": 650124
                },
                "children": [
                    {
                        "type": "MatchAllDocsQuery",
                        "description": "*:*",
                        "time_in_nanos": 345442099,
                        "breakdown": {
                            "set_min_competitive_score_count": 0,
                            "match_count": 0,
                            "shallow_advance_count": 0,
                            "set_min_competitive_score": 0,
                            "next_doc": 0,
                            "match": 0,
                            "next_doc_count": 0,
                            "score_count": 0,
                            "compute_max_score_count": 0,
                            "compute_max_score": 0,
                            "advance": 343032288,
                            "advance_count": 2279558,
                            "score": 0,
                            "build_scorer_count": 63,
                            "create_weight": 8618,
                            "shallow_advance": 0,
                            "create_weight_count": 1,
                            "build_scorer": 121571
                        }
                    },
                    {
                        "type": "LongDistanceFeatureQuery",
                        "description": "LongDistanceFeatureQuery(field=,origin=0,pivotDistance=787892250)",
                        "time_in_nanos": 777135963,
                        "breakdown": {
                            "set_min_competitive_score_count": 100,
                            "match_count": 0,
                            "shallow_advance_count": 63,
                            "set_min_competitive_score": 143873,
                            "next_doc": 0,
                            "match": 0,
                            "next_doc_count": 0,
                            "score_count": 2279537,
                            "compute_max_score_count": 42,
                            "compute_max_score": 6321,
                            "advance": 345943515,
                            "advance_count": 2279537,
                            "score": 426409120,
                            "build_scorer_count": 42,
                            "create_weight": 1816,
                            "shallow_advance": 9563,
                            "create_weight_count": 1,
                            "build_scorer": 62433
                        }
                    }
                ]
            }
        ],
        "rewrite_time": 53904,
        "collector": [
            {
                "name": "CancellableCollector",
                "reason": "search_cancelled",
                "time_in_nanos": 1217521280,
                "children": [
                    {
                        "name": "SimpleFieldCollector",
                        "reason": "search_top_hits",
                        "time_in_nanos": 556742784
                    }
                ]
            }
        ]
    }
]
```

Profile from one shard `desc`
```js
"searches": [
    {
        "query": [
            {
                "type": "BooleanQuery",
                "description": "#*:* LongDistanceFeatureQuery(field=,origin=1575784500,pivotDistance=787892250)",
                "time_in_nanos": 3533470,
                "breakdown": {
                    "set_min_competitive_score_count": 122,
                    "match_count": 0,
                    "shallow_advance_count": 0,
                    "set_min_competitive_score": 1167420,
                    "next_doc": 678521,
                    "match": 0,
                    "next_doc_count": 2844,
                    "score_count": 2844,
                    "compute_max_score_count": 0,
                    "compute_max_score": 0,
                    "advance": 22470,
                    "advance_count": 21,
                    "score": 1279540,
                    "build_scorer_count": 42,
                    "create_weight": 27929,
                    "shallow_advance": 0,
                    "create_weight_count": 1,
                    "build_scorer": 351716
                },
                "children": [
                    {
                        "type": "MatchAllDocsQuery",
                        "description": "*:*",
                        "time_in_nanos": 503232,
                        "breakdown": {
                            "set_min_competitive_score_count": 0,
                            "match_count": 0,
                            "shallow_advance_count": 0,
                            "set_min_competitive_score": 0,
                            "next_doc": 0,
                            "match": 0,
                            "next_doc_count": 0,
                            "score_count": 0,
                            "compute_max_score_count": 0,
                            "compute_max_score": 0,
                            "advance": 417693,
                            "advance_count": 3243,
                            "score": 0,
                            "build_scorer_count": 63,
                            "create_weight": 7558,
                            "shallow_advance": 0,
                            "create_weight_count": 1,
                            "build_scorer": 74674
                        }
                    },
                    {
                        "type": "LongDistanceFeatureQuery",
                        "description": "LongDistanceFeatureQuery(field=,origin=1575784500,pivotDistance=787892250)",
                        "time_in_nanos": 2072289,
                        "breakdown": {
                            "set_min_competitive_score_count": 122,
                            "match_count": 0,
                            "shallow_advance_count": 63,
                            "set_min_competitive_score": 1131340,
                            "next_doc": 0,
                            "match": 0,
                            "next_doc_count": 0,
                            "score_count": 2844,
                            "compute_max_score_count": 42,
                            "compute_max_score": 5662,
                            "advance": 396083,
                            "advance_count": 2863,
                            "score": 475990,
                            "build_scorer_count": 42,
                            "create_weight": 1228,
                            "shallow_advance": 6652,
                            "create_weight_count": 1,
                            "build_scorer": 49357
                        }
                    }
                ]
            }
        ],
        "rewrite_time": 43716,
        "collector": [
            {
                "name": "CancellableCollector",
                "reason": "search_cancelled",
                "time_in_nanos": 3154714,
                "children": [
                    {
                        "name": "SimpleFieldCollector",
                        "reason": "search_top_hits",
                        "time_in_nanos": 2412882
                    }
                ]
            }
        ]
    }
]
```

mayya-sharipova commented:
Results on the index `http_logs` with optimization:

desc:
```
|   50th percentile latency | desc_sort_timestamp |    399.83 |      ms |
|   90th percentile latency | desc_sort_timestamp |   429.145 |      ms |
|   99th percentile latency | desc_sort_timestamp |   459.862 |      ms |
|  100th percentile latency | desc_sort_timestamp |   464.219 |      ms |
```

asc:
```
|   50th percentile latency |  asc_sort_timestamp |   158.793 |      ms |
|   90th percentile latency |  asc_sort_timestamp |   170.675 |      ms |
|   99th percentile latency |  asc_sort_timestamp |    194.31 |      ms |
|  100th percentile latency |  asc_sort_timestamp |   200.831 |      ms |
```

mayya-sharipova commented:
I experimented why `LongDistanceFeatureQuery` doesn't not bring optimizations in the `asc` case on `population` on geonames. And it looks like `setMinCompetitiveScore` in this case is called, but because how  we use score calculation to compute `minValue` and `maxValue` for the range, we end up with broader ranges then necessary.  The cause of this is rounding errors in score calculation.

Several things to help to improve situation:
- In the `QueryPhase` for `pivotDistance` instead of average value for the field, use median value.  This help to avoid a situation of super small scores, where we loose some precision because of float calculation.
- If I set a proper median value for the `pivotDistance`, in the `LongDistanceFeatureQuery::minCompetiveScore` I will end up with a range of minValue=0 and maxValue = 0.  I wonder if we can exclude this type of ranges where minValue is equal to maxValue.

mayya-sharipova commented:
@jimczi @jpountz As agreed in our last conversation I have added an optional sort parameter `"optimized": true` that is `false` by default.  Can you please review this PR, so we can merge it and continue further iterations on top of it.

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -204,6 +220,29 @@ static boolean execute(SearchContext searchContext,
+                        searchContext.sort(new SortAndFormats(new Sort(newSortFields), newFormats));
+                    }
+                } catch (IOException e) {
+                    // in case of errors do nothing - keep the same query
```
this doesn't look safe as we might have changed the query already when this exception is caught?

mayya-sharipova replied:
@jpountz Can you please explain how this may happen? The way I see it is that the only way we can change the query if there is not exception in `tryRewriteLongSort`.

jpountz replied:
Imagine the case that the exception is thrown after `query = rewrittenQuery;`. Maybe there is nothing that throws an exception today, but even then this would be fragile and this invariant could get broken by a seemingly safe refactoring.

mayya-sharipova replied:
@jpountz Thanks, Adrien; makes sense. This addressed in the last commit, I have removed the `try` block.  I think it is reasonable if there are some IO exceptions during sort optimization, we can fail the whole search request.

jpountz commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -306,6 +352,92 @@ static boolean execute(SearchContext searchContext,
+             if (hasFilterCollector) return null;
+             // if we can't pre-calculate hitsCount based on the query type, optimization does't make sense
+             if (shortcutTotalHitCount(reader, query) == -1) return null;
+         }
```
the above lines seem to be indented by one more space than the lines below

mayya-sharipova commented:
@elasticmachine  run elasticsearch-ci/2

mayya-sharipova commented:
from now on, the work will continue of the feature branch https://github.com/elastic/elasticsearch/tree/long_sort_optimization
